aws-rds-dbinstance/src/main/java/software/amazon/rds/dbinstance/DeleteHandler.java

```diff
@@ -42,25 +46,23 @@ protected ProgressEvent<ResourceModel, CallbackContext> handleRequest(
             final VersionedProxyClient<RdsClient> rdsProxyClient,
             final VersionedProxyClient<Ec2Client> ec2ProxyClient
     ) {
+        // !!!: For delete handlers, the only input guaranteed to exist in the model is the DB instance identifier.
+        // This is why we fetch the DB instance instead of poking into the model.
+        // https://docs.aws.amazon.com/cloudformation-cli/latest/userguide/resource-type-test-contract.html#resource-type-test-contract-delete
         final ResourceModel resourceModel = request.getDesiredResourceState();
-        String snapshotIdentifier = null;
-        // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html
-        // For AWS::RDS::DBInstance resources that don't specify the DBClusterIdentifier property, the default policy is Snapshot.
-        // Final snapshots are not allowed for read replicas and cluster instances.
-        if (BooleanUtils.isNotFalse(request.getSnapshotRequested()) &&
-                !ResourceModelHelper.isReadReplica(resourceModel) &&
-                !isDBClusterMember(resourceModel)) {
-            snapshotIdentifier = snapshotIdentifierFactory.newIdentifier()
-                    .withStackId(request.getStackId())
-                    .withResourceId(StringUtils.prependIfMissing(request.getLogicalResourceIdentifier(), SNAPSHOT_PREFIX))
-                    .withRequestToken(request.getClientRequestToken())
-                    .toString();
-        }
-        final String finalSnapshotIdentifier = snapshotIdentifier;
 
         return ProgressEvent.progress(resourceModel, callbackContext)
+                .then(progress -> Commons.execOnce(progress, () -> {
+                        try {
+                            final var dbInstance = fetchDBInstance(rdsProxyClient.defaultClient(), progress.getResourceModel());
+                            callbackContext.setSnapshotIdentifier(decideSnapshotIdentifier(request, dbInstance));
+                        } catch (Exception exception) {
+                            return Commons.handleException(progress, exception, DEFAULT_DB_INSTANCE_ERROR_RULE_SET, requestLogger);
+                        }
+                        return progress;
+                }, CallbackContext::isDescribed, CallbackContext::setDescribed))
                 .then(progress -> proxy.initiate("rds::delete-db-instance", rdsProxyClient.defaultClient(), resourceModel, callbackContext)
-                        .translateToServiceRequest(model -> Translator.deleteDbInstanceRequest(model, finalSnapshotIdentifier))
+                        .translateToServiceRequest(model -> Translator.deleteDbInstanceRequest(model, callbackContext.getSnapshotIdentifier()))
```

Comment:
If the snapshot identifier is null, not a big fan of propagating it around the code base